nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/DistributedCacheClient.java

```diff
@@ -16,35 +16,23 @@
  */
 package org.apache.nifi.distributed.cache.client;
 
-import java.io.Closeable;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.util.concurrent.TimeUnit;
-
-import javax.net.ssl.SSLContext;
-
-public interface CommsSession extends Closeable {
-
-    void setTimeout(final long value, final TimeUnit timeUnit);
-
-    InputStream getInputStream() throws IOException;
-
-    OutputStream getOutputStream() throws IOException;
-
-    boolean isClosed();
+import org.apache.nifi.distributed.cache.client.adapter.InboundAdapter;
+import org.apache.nifi.distributed.cache.client.adapter.OutboundAdapter;
 
-    void interrupt();
-
-    String getHostname();
-
-    int getPort();
-
-    long getTimeout(TimeUnit timeUnit);
-
-    SSLContext getSSLContext();
-
-    int getProtocolVersion();
+import java.io.IOException;
 
-    void setProtocolVersion(final int protocolVersion);
+/**
+ * Encapsulate operations which may be performed using a {@link DistributedSetCacheClientService} or a
+ * {@link DistributedMapCacheClientService}.
+ */
+public interface DistributedCacheClient {
+
+    /**
+     * Call a service method.
+     *
+     * @param outboundAdapter the object used to assemble the service request byte stream
+     * @param inboundAdapter  the object used to interpret the service response byte stream
+     * @throws IOException on serialization failure; on communication failure
+     */
+    void invoke(OutboundAdapter outboundAdapter, InboundAdapter inboundAdapter) throws IOException;
```

Comment:
The is certainly a flexible interface, but it seems a bit too generic. What do you think about refactoring the interface to define the specific methods supported by the cache protocol? That would avoid requiring the caller to use string-based names for operations, but would provide a lower-level wrapper for various methods.  For example, a `put` method with a void return and byte array key and value. Essentially following the higher-level cache client interfaces, but closer to the protocol level.

Reply:
I changed this to a base class, and added implementations for Set & Map.  Also added enums to bound the supported operations.  It was a significant refactor, but I think the result is aligned with your suggestion.  LMK if this looks ok.